### PR TITLE
virtio_console.py: update and extend test for boot_to_much_ports

### PR DESCRIPTION
**Pull request by sitoliu**
(1) update and extend test for boot_to_much_ports
Port number has been extended from 32 to 512 for 2.x.x version, so update test case.

(2) disable loopback test between consoles
Data transfer test for console is invalid according the former test result.

(3) update error module with error_context
(4) remove try.. except in run() that covers every function
Without it step, exceptions can also be correctly raised in separate function, so delete this step to avoid importing avocado.core in this script.

id: 1472105, 1468463, 1468458

Signed-off-by: Sitong Liu <siliu@redhat.com>

### CODE REVIEW

File: qemu/tests/cfg/virtio_console.cfg [hunk 1]
```diff
@@ -227,11 +231,14 @@
...
-                            extra_params = " -device virtio-serial-pci,max_ports=32"
+                            extra_params_rhel = " -device virtio-serial-pci,max_ports=32"
+                            extra_params = " -device virtio-serial-pci,max_ports=512"
                             start_vm = no
                             virtio_console_test = failed_boot
-                            virtio_console_params = "maximum ports supported: 31"
+                            virtio_console_params_rhel = "maximum ports supported: 31"
+                            virtio_console_params = "maximum ports supported: 511"
                             virtio_ports = ""
+                            version_params = "1\.\d\.\d"
```
Comment (hereischen) [inline qemu/tests/cfg/virtio_console.cfg]: I'd suggest change `version_params` to `qemu_verison_patten`.
I am not sure if use version number is the best way. 
I saw you use `qemu_binary = utils_misc.get_qemu_binary(params)` below,
based on what I know rhev6 is qemu－kvm－rehv 0.x.x

Are you trying to make it compatible for rhev7 and rehl7 ?

File: qemu/tests/cfg/virtio_console.cfg [hunk 2]
```diff
@@ -227,11 +231,14 @@
                         - boot_too_much_ports:
                             # arm doesn't use pci bus
                             no aarch64
-                            extra_params = " -device virtio-serial-pci,max_ports=32"
+                            extra_params_rhel = " -device virtio-serial-pci,max_ports=32"
+                            extra_params = " -device virtio-serial-pci,max_ports=512"
```
Comment (hereischen) [inline qemu/tests/cfg/virtio_console.cfg]: I think you can use same param here, eg `extra_params = " -device virtio-serial-pci,max_ports=%s"`, then change the size accordingly.

File: qemu/tests/virtio_console.py [hunk 3]
```diff
@@ -1845,24 +1826,10 @@ def test_delete_guest_script():
...
-                                  exc_type, exc_value,
-                                  exc_traceback.tb_next)))
-            if isinstance(details, error.TestError):
-                raise error.TestError('%s error: %s' % (_fce, details))
-            elif isinstance(details, error.TestNAError):
-                raise error.TestNAError('%s skipped: %s' % (_fce, details))
-            else:
-                raise error.TestFail('%s failed: %s' % (_fce, details))
+        fce = locals()[_fce]
+        return fce()
```
Comment (hereischen) [inline qemu/tests/virtio_console.py]: why do you delete the `excpt` part?

File: qemu/tests/cfg/virtio_console.cfg (hunk 1 above)
Comment (hereischen) [inline qemu/tests/cfg/virtio_console.cfg]: I'd suggest change `version_params` to `qemu_verison_patten`.
I am not sure if use version number is the best way. 
I saw you use `qemu_binary = utils_misc.get_qemu_binary(params)` below,
based on what I know rhev6 is qemu－kvm－rehv 0.x.x

Are you trying to make it compatible for rhev7 and rehl7 ?

Comment (hereischen) [inline qemu/tests/cfg/virtio_console.cfg]: I'd suggest change `version_params` to `qemu_verison_patten`.
I am not sure if use version number is the best way. 
I saw you use `qemu_binary = utils_misc.get_qemu_binary(params)` below,
based on what I know rhev6 is qemu－kvm－rehv 0.x.x

Are you trying to make it compatible for rhev7 and rehl7 ?

File: qemu/tests/cfg/virtio_console.cfg (hunk 2 above)
Comment (sitoliu) [reply qemu/tests/cfg/virtio_console.cfg]: Thanks for advice. It's a good idea. I will update it.

File: qemu/tests/cfg/virtio_console.cfg (hunk 1 above)
Comment (sitoliu) [reply qemu/tests/cfg/virtio_console.cfg]: Yes, I am trying to make it compatible for rhev7 and rhel7, and rhel6 if possible, since the "max_ports" now is only extended on rhev7 version, I think if we match the version pattern with rhev7, the lower pattern will be well covered.

File: qemu/tests/virtio_console.py (hunk 3 above)
Comment (sitoliu) [reply qemu/tests/virtio_console.py]: Thanks. This code block applied `isinstance` to determine the exception type, which needs to import avocado.core module, so the reason is to avoid this importing in special test script.

**Comment by sitoliu** [comment]
RHEL7:
smp_2.128m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.virtio_console.spread_linear.incorrect_options.boot_too_much_ports: PASS (8.00 s)
RHEV7:
smp_2.128m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.virtio_console.spread_linear.incorrect_options.boot_too_much_ports: PASS (8.00 s)

Hi @hereischen, I updated my code, could you help review it again? Thanks!

File: qemu/tests/virtio_console.py [hunk 4]
```diff
@@ -1792,19 +1765,31 @@ def test_unplugged_restart():
...
 
-    @error.context_aware
+    @error_context.context_aware
     def test_failed_boot():
         """
         Start VM and check if it failed with the right error message.
         :param cfg: virtio_console_params - Expected error message.
         """
-        exp_error_message = params['virtio_console_params']
+        exp_error_message = (params['virtio_console_params'] % 31)
```
Comment (hereischen) [inline qemu/tests/virtio_console.py]: I think it may be better put these numbers（32，31）（512，511） into cfg and give a meaningful name. 
What about put this line into the else clause under line 1791? are they always minus one? 
And I noticed doc string of this function is outdated, can you please also update it please?

Comment (sitoliu) [reply qemu/tests/virtio_console.py]: Thanks for advice. I separated these nums to valid set and invalid set. (512, 32) are the minimum invalid boundary cases for `max_ports`, so they always be the valid `max_ports` + 1. 
I updated my code again and could you help review it the third time? Thanks!

RHEV7:
smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.virtio_console.spread_linear.incorrect_options.boot_too_much_ports: PASS (6.47 s)
RHEL7:
smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.virtio_console.spread_linear.incorrect_options.boot_too_much_ports: PASS (6.16 s)

File: qemu/tests/virtio_console.py [hunk 5]
```diff
@@ -1792,19 +1765,40 @@ def test_unplugged_restart():
...
         Start VM and check if it failed with the right error message.
         :param cfg: virtio_console_params - Expected error message.
-        """
-        exp_error_message = params['virtio_console_params']
+        :param cfg: qemu_version_pattern - Qemu version pattern for high version.
+        :param cfg: max_ports_invalid - Invalid max ports nums for different versions.
+        :param cfg: max_ports_valid - Valid max ports nums for different versions.
+
+        """
+        max_ports_invalid = params['max_ports_invalid']
```
Comment (hereischen) [inline qemu/tests/virtio_console.py]: how about put these in one line `max_ports_invalid = params['max_ports_invalid'].split(',')`

**Comment by sitoliu** [comment]
@hereischen Thanks, updated the code.

RHEL7:
smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.virtio_console.spread_linear.incorrect_options.boot_too_much_ports: PASS (6.15 s)
RHEV7:
smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.virtio_console.spread_linear.incorrect_options.boot_too_much_ports: PASS (6.18 s)

File: qemu/tests/virtio_console.py (hunk 3 above)
Comment (luckyh) [reply qemu/tests/virtio_console.py]: Seems we should keep `EXIT_EVENT.set()`.

Comment (sitoliu) [reply qemu/tests/virtio_console.py]: @luckyh Thank for advice, I modified my code, could you help review it again? Thanks!

RHEL7:
smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.virtio_console.spread_linear.incorrect_options.boot_too_much_ports: PASS (6.05 s)
RHEV7:
smp_2.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.myprovider.virtio_console.spread_linear.incorrect_options.boot_too_much_ports: PASS (6.09 s)

**Comment by luckyh** [comment]
Thanks all.
